### ISSUE TITLE
[Q] Do not write segment if specific value is not available

### ISSUE THREAD

**Issue by niravbhattsai** [indice-co/EDI.Net#136]
Hello,

First of all, thank you very much for this project. This works very well for me.

I am trying to create a segment as below. 

```csharp
    [EdiSegment, EdiPath("DTP")]
    public class DTP
    {
        [EdiValue("X(3)", Path = "DTP/0", Description = "DTP01 - Date/Time Qualifier")]
        public string DateTimeQualifier { get; set; }

        [EdiValue("X(3)", Path = "DTP/1", Description = "DTP02 - Date Time Period Format Qualifier")]
        public string DateTimePeriodFormatQualifier { get; set; }

        [EdiValue("X(35)", Path = "DTP/2", Description = "DTP03 - Date Time Period")]
        public string DateTimePeriod { get; set; }

        public static DTP GetDTPyyyyMMdd(string qualifier, DateTime? date1, DateTime? date2= null)
        {
            string datePeriod=null, periodFormatQualifier=null;
            if(date1.HasValue && date2.HasValue)
            {
                periodFormatQualifier = "RD8";
                datePeriod = $"{date1:yyyyMMdd}-{date2:yyyyMMdd}";
            }
            else if(date1.HasValue)
            {
                periodFormatQualifier = "D8";
                datePeriod = $"{date1:yyyyMMdd}";
            }
            if(string.IsNullOrWhiteSpace(datePeriod))
            {
                return null;
            }
            else
            {
                return new DTP() {DateTimeQualifier= qualifier, DateTimePeriodFormatQualifier=periodFormatQualifier, DateTimePeriod=datePeriod };
            }
        }

        public static DTP GetDTPCCyyMMdd(string qualifier, DateTime? date1, DateTime? date2 = null)
        {
            string datePeriod=null, periodFormatQualifier=null;
            if (date1.HasValue && date2.HasValue)
            {
                periodFormatQualifier = "RD8";
                datePeriod = $"{(date1.Value.Year/100)+1}{date1:yyMMdd}-{(date2.Value.Year / 100) + 1}{date2:yyMMdd}";
            }
            else if (date1.HasValue)
            {
                periodFormatQualifier = "D8";
                datePeriod = $"{(date1.Value.Year / 100) + 1}{date1:yyMMdd}";
            }
            if (string.IsNullOrWhiteSpace(datePeriod))
            {
                return null;
            }
            else
            {
                return new DTP() { DateTimeQualifier = qualifier, DateTimePeriodFormatQualifier = periodFormatQualifier, DateTimePeriod = datePeriod };
            }
        }
    }
```

Now I don't want to generate DTP segment if my date1 is null.
Here I have provided null in the if block, but I have tried with providing just new DTP() as well.

For example 
```csharp
DTP dtpObject = DTP.GetDTPyyyyMMdd("472", _date1);
            StringBuilder sb = new StringBuilder();
            using (var textWriter = new StringWriter(sb))
            {
                using (var ediWriter = new EdiTextWriter(textWriter, grammar))
                {
                    new EdiSerializer().Serialize(ediWriter, dtpObject);
                }
            }
            string dtpOutput = sb.ToString();
```

It generates correct output **DTP\*472\*D8\*20190815~**

When I provide null as date1 my object will either null or an empty object. 
When the object is null, serialize will raise the exception that token is null.
When the object is new DTP(), serialize will generate output **DTP~**. 

I do not want to generate the segment if my date period is null. 

I have tried with EDIValue.Mandatory flag, but it gives same output **DTP~**.

Now when I put DTP in any other segment is null, it does not give any error as well as it does generate the value. 
Now if the value is at the top level, is it possible not to generate the segment?

When EDI is written it always add trailing "\r\n", is there any way to remove that?

**Comment by niravbhattsai** [comment]
Another question is, how do you suggest calculating SE01 - Number of included segments.
As the segments will be generated at run time.

**Comment by cleftheris** [comment]
Hi and thanks for your interest in the library. I am currently on vacations and won’t be able to see your issue until September. 

Ps The segment count is something that I am planning to support autogeneration. Nothing concrete yet.

**Comment by niravbhattsai** [comment]
I found another complex scenario. Let me know if I am doing something wrong here or its a different issue.
[IndiceTrial.txt](https://github.com/indice-co/EDI.Net/files/3563714/IndiceTrial.txt)



Please find attached class file.

I am trying to generate an EDI segment like
HI*ABK:1*ABK:2*ABK:3.....*ABK:12~

But it is being generated as
HI*ABK:1~
HI*ABK:2~ 
..
HI*ABK:12~

I want to use multiple instances of the same Edi Element.

**Comment by cleftheris** [comment]
Hi @niravbhattsai. This is a very nice test! Actually I have started working on reusing element classes some months ago but only got partial support. 

In order for this to work as intended you should make use of wildcards in your paths inside the `HealthCareCodeInfo` class like so:

```csharp
        [EdiElement, EdiPath("HI/*")]
        public class HealthCareCodeInfo
        {
            [EdiValue("X(3)", Path ="HI/*/0")]
            public string Qualifier { get; set; }
            [EdiValue("X(10)", Path = "HI/*/1")]
            public string Code { get; set; }
        }
```

I will add a test case with this scenario and check it out.

**Comment by cleftheris** [comment]
One more note here. This is not meant to be an endless discussion on various issues you may encounter. In order for this to be helpfull to anyone else you must break down your questions to different issues with one clear exaple of what you tried and what is the expected behavior. 

Otherwize this issue will never be able to be addressed and closed

**Comment by cleftheris** [comment]
> I found another complex scenario. Let me know if I am doing something wrong here or its a different issue.
> [IndiceTrial.txt](https://github.com/indice-co/EDI.Net/files/3563714/IndiceTrial.txt)
> 
> Please find attached class file.
> 
> I am trying to generate an EDI segment like
> HI_ABK:1_ABK:2*ABK:3.....*ABK:12~
> 
> But it is being generated as
> HI_ABK:1~ HI_ABK:2~
> ..
> HI*ABK:12~
> 
> I want to use multiple instances of the same Edi Element.

Regarding your question here I have just put up a release that fixes this 1.9.4.

Also if you want an segment to be omitted from the serialization you just make the property that represents the whole segment null.
